### DISCUSSION TITLE
`JsonGenerator` supporting `JsonIdentityInfo` annotation

### DISCUSSION THREAD

**Discussion by knutwannheden** [FasterXML/jackson-databind#4052] [Q&A]
I have a long array of objects, where the objects are deeply nested and inter-connected object graphs (also with cycles) and an object owned by one root in the array may also reference objects owned by another root. To make sure objects only get serialized once, I am using Jackson's `@JsonIdentityInfo` annotation which works very nicely.

Now, since the array can get very large and is streamed from an external source into memory, I would like to also serialize these objects as a stream, rather than load everything into a `List` and then serialize that. I thought this may be a good use case for `JsonGenerator`. When I write the first object using `JsonGenerator#writeObject()` that works nicely, but when I then write the second object, any references to the first object (or an object owned by it) don't get serialized as references. Instead the IDs are reset to 1.

So now to my question: Should I:
- register my own `DefaultSerializerProvider` implementation which in `createInstance()` always returns the same instance? (I don't see any way for it to return an instance specific to the `JsonGenerator` instance, which makes me think this is not such a good idea.)
- Create a POJO to represent the whole array, which somehow implements the streaming when Jackson serializes it. (Not sure if this is possible or how it would be done.)
- Another solution I don't know of?

Thanks for your help!

**Comment by knutwannheden** [comment]
I tried the second approach with a custom serializer. The streaming part worked well (in my POJO to be serialized the stream of objects is represented as a typed `Stream`), but I had the same problem there: When an object is encountered again within a new `JsonGenerator#writeObject()` call, it will again be serialized, rather than just an ID reference.

My assumption was that the outermost serialization invocation would somehow initialize some kind of "transaction context" within which the IDs would be assigned to the objects being serialized and that when the custom serializer would operate within this transaction, it would automatically serialize objects as references, if they had already previously been serialized.

**Reply by knutwannheden** [reply]
Any ideas, @JooHyukKim or @cowtowncoder ? Please let me know if you would like me to amend my question with some code to better illustrate the problem.

**Reply by JooHyukKim** [reply]
@knutwannheden hmmm, though I need to read this through after work (in about 7 hours 😅), I think similar to the comment below https://github.com/FasterXML/jackson-databind/discussions/4052#discussioncomment-6546141, that to make use of the full feature, should work at the same or higher(more abstracted-away) level such as `JsonDe/serializer` (also mentioned at the comment).

**Reply by cowtowncoder** [reply]
As per my other answer, the context (`SerializerProvider` instance) only exists when using methods at databind level; with Jackson 2.x it is not propagated through `JsonGenerator`. That is why `JsonGenerator` write methods do not have it.
So `JsonGenerator` methods are only to be used for actual output (token) generation.

**Reply by knutwannheden** [reply]
Thank you both for taking your time. I will try to see if I can somehow get this to work using a `JsonSerializer` instead. My readon for using `JsonGenerator` was that I wanted to serialize my `Stream<MyObject>` by first writing an array-start token, then a list of objects, finally an array-end token and the `JsonGenerator` looked like a convenient way to achieve this. But I will see how far I get with `JsonSerializer` and report back.

**Comment by cowtowncoder** [comment]
Just one quick note (alas, I don't have time to dig very deep): none of `writeXxx()` methods in `JsonGenerator` should be used, in general, as they are likely to NOT handled "advanced" features like Object Id or Type Id handling. Writes by serializers need to operate at databind level, by delegating to resolved `JsonSerializer`s instead.

So, `JsonGenerator.writeObject()` in particular is almost guaranteed not to work with Object Ids, do not (try to) use it.
Part of this has to do with context (SerializerProvider) construction; that can't be fixed for 2.x. 3.0 (master) has better handling, but even there I wouldn't recommend going via `JsonGenerator`.
